defi/src/api2/scripts/_test2.ts

```diff
@@ -0,0 +1,197 @@
+  const latestProtocolItemsMonthAgoMap: any = {}
+  latestProtocolItemsMonthAgo.forEach((data: any) => latestProtocolItemsMonthAgoMap[data.id] = data.data)
+  console.timeEnd('getLatestProtocolItems filterAMonthAgo')
+
+  console.time('getAllProtocolItems')
+  const protocolDataMap: { [key: string]: any } = {}
+  metadata.protocolDataMap = protocolDataMap
+
+  let i = 0
+  await PromisePool.withConcurrency(20)
+    .for(protocols)
+    .process(async (protocol: any) => {
+      try {
+        const dataObj: any = {}
+        protocolDataMap[protocol.id] = dataObj
+        const tvlData = await getAllProtocolItems(dailyTvl, protocol.id)
+
+        const hourlyItem = latestProtocolItemsMap[protocol.id]
+        dataObj.tvlData = tvlData
+        dataObj.lastHourlyRecord = hourlyItem
+        dataObj.tvlADayAgo = latestProtocolItemsDayAgoMap[protocol.id]
+        dataObj.tvlAWeekAgo = latestProtocolItemsWeekAgoMap[protocol.id]
+        dataObj.tvlAMonthAgo = latestProtocolItemsMonthAgoMap[protocol.id]
+
+        // if (hourlyItem) {
+        //   if (!tvlData.length || tvlData[tvlData.length - 1].SK < hourlyItem.SK)
+        //     tvlData.push(hourlyItem)
+        // }
+        i++
+        if (i % 100 === 0) log(i, 'tvlData.length', tvlData.length, !!hourlyItem, protocol.name)
+      } catch (e) {
+        console.error(e)
+      }
+    });
+  console.timeEnd('getAllProtocolItems')
+
+
+  /* console.time('allTvlData')
+  let limit = 100000
+  let offset = 0
+  const allTvlData: any[] = []
+  let lastLength
+  do {
+    lastLength = allTvlData.length
+    const res = await TABLES.DAILY_TVL.findAll({ limit, offset, raw: true, attributes: ['id', 'data', 'timestamp'] })
+    allTvlData.push(...res)
+    offset += limit
+    log('allTvlData.length', allTvlData.length)
+  } while (allTvlData.length % limit === 0 && allTvlData.length !== lastLength)
+  console.timeEnd('allTvlData')
+ */
+
+  /*   console.time('allTvlDataMap')
+    const protocolMap: any = {}
+    protocols.forEach((protocol: any) => protocolMap[protocol.id] = protocol)
+    allTvlData.forEach(({ id, data, timestamp }: any) => {
+      data.SK = timestamp
+      const protocol = protocolMap[id]
+      if (protocol) {
+        if (!Array.isArray(protocol.tvlData)) protocol.tvlData = []
+        protocol.tvlData.push(data)
+      }
+    })
+    protocols.forEach((protocol: any) => {
+      if (Array.isArray(protocol.tvlData))
+        protocol.tvlData.sort((a: any, b: any) => a.SK - b.SK)
+      const hourlyItem = latestProtocolItemsMap[protocol.id]
+      if (hourlyItem) {
+        if (!protocol.tvlData.length || protocol.tvlData[protocol.tvlData.length - 1].SK !== hourlyItem.SK)
+          protocol.tvlData.push(hourlyItem)
+      }
+      if (!Array.isArray(protocol.tvlData)) protocol.tvlData = []
+  
+    })
+    console.timeEnd('allTvlDataMap') */
+
+
+  console.time('write data')
+  fs.writeFileSync('../../allTvlData.json', JSON.stringify(metadata))
```

Comment:
should set up a persisting disk volume maybe

Reply:
you can ignore this code, what I did here was save to temp, so when this test script is run we dont query db each time instead use temp/test data.


In the actual script, we will read from db, keep it in memory and use it for all computation done